File: src/CircleAI.Gaming/GamingPrimitives.cs
// GamingPrimitives.cs — (3.3.0)
using System;
using System.Collections.Concurrent;
using System.Collections.Generic;
using System.Linq;

namespace CircleAI.Gaming;

public sealed record GameTitle(string TitleId, string Name, string Genre, string Platform);
public sealed record PlaySession(string SessionId, string UserId, string TitleId, TimeSpan Duration, DateTimeOffset AtUtc);
public sealed record AchievementUnlock(string UnlockId, string UserId, string TitleId, string Achievement, DateTimeOffset AtUtc);

public interface IGamingBoard
{
    void AddTitle(GameTitle t);
    GameTitle? GetTitle(string id);
    IReadOnlyList<GameTitle> TitlesByGenre(string genre);
    void RecordSession(PlaySession s);
    TimeSpan TotalPlayTime(string userId, string titleId);
    void Unlock(AchievementUnlock u);
    IReadOnlyList<AchievementUnlock> AchievementsFor(string userId);
    IReadOnlyList<GameTitle> MostPlayed(string userId, int topK = 5);
}

public sealed class InMemoryGamingBoard : IGamingBoard
{
    private readonly ConcurrentDictionary<string, GameTitle> _titles = new(StringComparer.Ordinal);
    private readonly List<PlaySession> _sessions = new();
    private readonly List<AchievementUnlock> _unlocks = new();
    private readonly object _lock = new();

    public void AddTitle(GameTitle t) { ArgumentNullException.ThrowIfNull(t); _titles[t.TitleId] = t; }
    public GameTitle? GetTitle(string id) => _titles.GetValueOrDefault(id);
    public IReadOnlyList<GameTitle> TitlesByGenre(string genre)
        => _titles.Values.Where(t => string.Equals(t.Genre, genre, StringComparison.OrdinalIgnoreCase)).ToArray();
    public void RecordSession(PlaySession s) { ArgumentNullException.ThrowIfNull(s); lock (_lock) _sessions.Add(s); }
    public TimeSpan TotalPlayTime(string userId, string titleId)
    {
        lock (_lock)
        {
            var ms = _sessions.Where(s => s.UserId == userId && s.TitleId == titleId).Sum(s => s.Duration.TotalMilliseconds);
            return TimeSpan.FromMilliseconds(ms);
        }
    }
    public void Unlock(AchievementUnlock u) { ArgumentNullException.ThrowIfNull(u); lock (_lock) _unlocks.Add(u); }
    public IReadOnlyList<AchievementUnlock> AchievementsFor(string userId)
    { lock (_lock) return _unlocks.Where(u => u.UserId == userId).OrderByDescending(u => u.AtUtc).ToArray(); }
    public IReadOnlyList<GameTitle> MostPlayed(string userId, int topK = 5)
    {
        if (topK <= 0) throw new ArgumentOutOfRangeException(nameof(topK));
        lock (_lock)
        {
            return _sessions.Where(s => s.UserId == userId)
                .GroupBy(s => s.TitleId)
                .OrderByDescending(g => g.Sum(s => s.Duration.TotalMilliseconds))
                .Take(topK)
                .Select(g => _titles.TryGetValue(g.Key, out var t) ? t : null)
                .Where(t => t is not null)
                .Cast<GameTitle>()
                .ToArray();
        }
    }
}

File: python/src/circle_ai/gaming/gaming_primitives.py
# gaming_primitives.py
#
# Port of CircleAI.Gaming GamingPrimitives.cs (C# — the EXACT spec).
#
# (3.3.0) Real domain types + in-memory board for the Gaming vertical: game
# titles, play sessions, achievement unlocks. C# ConcurrentDictionary -> dict;
# the session/unlock lists are guarded by a single lock. TimeSpan -> timedelta,
# DateTimeOffset -> datetime. TotalPlayTime sums session durations; MostPlayed
# groups sessions by title, orders by total play time desc, and maps to titles.

from __future__ import annotations

import threading
from abc import ABC, abstractmethod
from collections import defaultdict
from dataclasses import dataclass
from datetime import datetime, timedelta
from typing import Dict, List, Optional


@dataclass(frozen=True, slots=True)
class GameTitle:
    """Mirrors ``CircleAI.Gaming.GameTitle``."""

    title_id: str
    name: str
    genre: str
    platform: str


@dataclass(frozen=True, slots=True)
class PlaySession:
    """Mirrors ``CircleAI.Gaming.PlaySession``."""

    session_id: str
    user_id: str
    title_id: str
    duration: timedelta
    at_utc: datetime


@dataclass(frozen=True, slots=True)
class AchievementUnlock:
    """Mirrors ``CircleAI.Gaming.AchievementUnlock``."""

    unlock_id: str
    user_id: str
    title_id: str
    achievement: str
    at_utc: datetime


class IGamingBoard(ABC):
    """In-memory board for game titles, play sessions and achievements."""

    @abstractmethod
    def add_title(self, t: GameTitle) -> None:
        ...

    @abstractmethod
    def get_title(self, id: str) -> Optional[GameTitle]:
        ...

    @abstractmethod
    def titles_by_genre(self, genre: str) -> List[GameTitle]:
        ...

    @abstractmethod
    def record_session(self, s: PlaySession) -> None:
        ...

    @abstractmethod
    def total_play_time(self, user_id: str, title_id: str) -> timedelta:
        ...

    @abstractmethod
    def unlock(self, u: AchievementUnlock) -> None:
        ...

    @abstractmethod
    def achievements_for(self, user_id: str) -> List[AchievementUnlock]:
        ...

    @abstractmethod
    def most_played(self, user_id: str, top_k: int = 5) -> List[GameTitle]:
        ...


class InMemoryGamingBoard(IGamingBoard):
    """Thread-safe in-memory :class:`IGamingBoard`."""

    def __init__(self) -> None:
        self._titles: Dict[str, GameTitle] = {}
        self._sessions: List[PlaySession] = []
        self._unlocks: List[AchievementUnlock] = []
        self._lock = threading.Lock()

    def add_title(self, t: GameTitle) -> None:
        if t is None:
            raise ValueError("game title must not be None")
        with self._lock:
            self._titles[t.title_id] = t

    def get_title(self, id: str) -> Optional[GameTitle]:
        with self._lock:
            return self._titles.get(id)

    def titles_by_genre(self, genre: str) -> List[GameTitle]:
        target = genre.casefold()
        with self._lock:
            return [
                t for t in self._titles.values() if t.genre.casefold() == target
            ]

    def record_session(self, s: PlaySession) -> None:
        if s is None:
            raise ValueError("play session must not be None")
        with self._lock:
            self._sessions.append(s)

    def total_play_time(self, user_id: str, title_id: str) -> timedelta:
        with self._lock:
            total = timedelta()
            for s in self._sessions:
                if s.user_id == user_id and s.title_id == title_id:
                    total += s.duration
            return total

    def unlock(self, u: AchievementUnlock) -> None:
        if u is None:
            raise ValueError("achievement unlock must not be None")
        with self._lock:
            self._unlocks.append(u)

    def achievements_for(self, user_id: str) -> List[AchievementUnlock]:
        with self._lock:
            items = [u for u in self._unlocks if u.user_id == user_id]
        items.sort(key=lambda u: u.at_utc, reverse=True)
        return items

    def most_played(self, user_id: str, top_k: int = 5) -> List[GameTitle]:
        if top_k <= 0:
            raise ValueError("top_k must be positive")
        with self._lock:
            totals: Dict[str, timedelta] = defaultdict(timedelta)
            for s in self._sessions:
                if s.user_id == user_id:
                    totals[s.title_id] += s.duration
            # OrderByDescending(total).Take(topK), then map title ids to titles,
            # dropping any that are missing (C# Where(t => t is not null)).
            ranked = sorted(
                totals.items(), key=lambda kv: kv[1].total_seconds(), reverse=True
            )[:top_k]
            result: List[GameTitle] = []
            for title_id, _ in ranked:
                t = self._titles.get(title_id)
                if t is not None:
                    result.append(t)
            return result
